tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp

```diff
@@ -0,0 +1,297 @@
+
+  std::array<DataVector, 4> char_speeds_one{
+      {get(one), get(one), get(one), get(one)}};
+  std::array<DataVector, 4> char_speeds_minus_one{
+      {get(minus_one), get(minus_one), get(minus_one), get(minus_one)}};
+  std::array<DataVector, 4> char_speeds_five{
+      {get(five), get(five), get(five), get(five)}};
+  std::array<DataVector, 4> char_speeds_minus_five{
+      {get(minus_five), get(minus_five), get(minus_five), get(minus_five)}};
+
+  GeneralizedHarmonic::UpwindFlux<spatial_dim> flux_computer{};
+
+  // If all the char speeds are +1, the weighted fields should just
+  // be the interior fields
+  const auto weighted_char_fields_one =
+      GeneralizedHarmonic::GeneralizedHarmonic_detail::weight_char_fields<3>(
+          char_fields_int, char_speeds_one, char_fields_ext, char_speeds_one);
+  CHECK(weighted_char_fields_one == char_fields_int);
+
+  // If all the char speeds are -1, the weighted fields should just be
+  // the exterior fields up to a sign
+  const auto weighted_char_fields_minus_one =
+      GeneralizedHarmonic::GeneralizedHarmonic_detail::weight_char_fields<3>(
+          char_fields_int, char_speeds_minus_one, char_fields_ext,
+          char_speeds_minus_one);
+  CHECK(weighted_char_fields_minus_one == -1.0 * char_fields_ext);
+
+  // Check scaling by 5 instead of 1
+  const auto weighted_char_fields_minus_five =
+      GeneralizedHarmonic::GeneralizedHarmonic_detail::weight_char_fields<3>(
+          char_fields_int, char_speeds_minus_five, char_fields_ext,
+          char_speeds_minus_five);
+  const auto weighted_char_fields_five =
+      GeneralizedHarmonic::GeneralizedHarmonic_detail::weight_char_fields<3>(
+          char_fields_int, char_speeds_five, char_fields_ext, char_speeds_five);
+  CHECK(weighted_char_fields_minus_five == -5.0 * char_fields_ext);
+  CHECK(weighted_char_fields_five == 5.0 * char_fields_int);
+
+  // Consistency checks
+  Variables<typename GeneralizedHarmonic::UpwindFlux<spatial_dim>::package_tags>
+  packaged_data_int(get(one).size(),
+                    std::numeric_limits<double>::signaling_NaN());
+  flux_computer.package_data(make_not_null(&packaged_data_int),
+                             spacetime_metric_int, pi_int, phi_int, lapse_int,
+                             shift_int, inverse_spatial_metric_int, gamma_1,
+                             gamma_2, unit_normal_one_form_int);
+  Variables<typename GeneralizedHarmonic::UpwindFlux<spatial_dim>::package_tags>
+  packaged_data_ext(get(one).size(),
+                    std::numeric_limits<double>::signaling_NaN());
+  flux_computer.package_data(make_not_null(&packaged_data_ext),
+                             spacetime_metric_ext, pi_ext, phi_ext, lapse_ext,
+                             shift_ext, inverse_spatial_metric_ext, gamma_1,
+                             gamma_2, unit_normal_one_form_ext);
+
+  const auto minus_unit_normal_one_form_int =
+      divide_by(unit_normal_one_form_int, get(minus_one));
```

Comment:
I'm fine changing this, but why is this abuse? I would have thought this is fine, with the advantage of letting divide_by handle the looping to keep the code here simpler.

Also, why does spectre provide divide_by() if it's not used anywhere and should not be used in a case like this? When *should* divide_by be used?

Reply:
I wouldn't say it's so much of an abuse as it is just writing the code in a rather inefficient and clumsy manner... (though I can definitely see how this can be considered abuse :) ) You basically do a copy and create a datavector of all `-1` instead of doing a single for loop.

Looking at the implementation of `divide_by` it's not really clear to me what the intended use case is or why it exists at all. @osheamonn added it so hopefully he can answer these questions.

Reply:
We should probably add simple tensor arithmetic asap (as soon as someone finds the time and energy to do that) to support what a `Variables` can do already, i.e. addition and scalar multiplication at least...

Reply:
In my latest commit, I removed the divide_by's, so this shouldn't hold up this PR. But I am curious what the result of the divide_by discussion is.

Reply:
`divide_by` was ported from bitbucket extremely early on in the porting process, when we were trying to port the small and easy things first. I presume that wherever it was used in bitbucket (probably to normalise unit vectors is my guess) didn't end up being ported, and that ended up being handled by the Interface items infrastructure which was added almost a year later. So yeah `divide_by` is probably not needed, but at the time it was added we thought it would be.